core/modules/replica/IndexJob.cc

```diff
@@ -0,0 +1,538 @@
+            }
+        }
+        if (worker.empty()) {
+            throw logic_error(context() + string(__func__) + ":  internal bug");
+        }
+        _chunks[worker].push_back(chunk);
+    }
+
+    // Launch the initial batch of requests in the number which won't exceed
+    // the number of the service processing threads at each worker multiplied
+    // by the number of workers involved into the operation and by the "magic"
+    // number 8. The later is needed to absorb the latency of the network
+    // communications so that the worker threads would be able to work on
+    // another batch of the data extraction requests while results of the
+    // previous batch were being sent back to the Controller.
+
+    size_t const maxRequestsPerWorker =
+        8 * controller()->serviceProvider()->config()->workerNumProcessingThreads();
+
+    for (auto&& worker: workerNames) {
+        for (auto&& ptr: _launchRequests(lock, worker, maxRequestsPerWorker)) {
+            _requests[ptr->id()] = ptr;
+        }
+    }
+
+    // In case if no workers or database are present in the Configuration
+    // at this time.
+
+    if (_requests.size() == 0) finish(lock, ExtendedState::SUCCESS);
+}
+
+
+void IndexJob::cancelImpl(util::Lock const& lock) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << __func__);
+
+    // The algorithm will also clear resources taken by various
+    // locally created objects.
+
+    _chunks.clear();
+
+    // To ensure no lingering "side effects" will be left after cancelling this
+    // job the request cancellation should be also followed (where it makes a sense)
+    // by stopping the request at corresponding worker service.
+
+    for (auto&& itr: _requests) {
+        auto&& ptr = itr.second;
+        ptr->cancel();
+        if (ptr->state() != Request::State::FINISHED) {
+            controller()->stopById<StopIndexRequest>(
+                ptr->worker(),
+                ptr->id(),
+                nullptr,    /* onFinish */
+                options(lock).priority,
+                true,       /* keepTracking */
+                id()        /* jobId */
+            );
+        }
+    }
+    _requests.clear();
+    _rollbackTransaction(__func__);
+}
+
+
+void IndexJob::notify(util::Lock const& lock) {
+    LOGS(_log, LOG_LVL_DEBUG, context() << __func__);
+    notifyDefaultImpl<IndexJob>(lock, _onFinish);
+}
+
+
+void IndexJob::_onRequestFinish(IndexRequest::Ptr const& request) {
+
+    // NOTE: this algorithm assumes "zero tolerance" to failures - any failure
+    // in executing requests or processing data of the requests would result in
+    // the job termination. The only exception from this rule is a scenario
+    // when a target chunk table won't have a partition. This may be expected
+    // for some chunk tables because they may not have contributions in a context
+    // of the given super-transaction.
+    //
+    // TODO: reconsider this algorithm.
```

Comment:
Chances are this may actually happen. This code is still in the "under development" case. It may need improvements later in some later iteration of the Ingest system. The current implementation is not quite tolerant for failures to harvest chunk-specific contributions into the "secondary index". Presently, I don't have any significant statistics on a probability of failures in this application. There were no problems with the current implementation on the large-scale tests (PDAC) tough. So, I put `TODO` as a flag indicating an area where further attention is needed.